### ISSUE TITLE
Testing assert/required using ThrowProxy

### ISSUE THREAD

**Issue by marcosmartinez7** [trufflesuite/truffle#1001]
Hi, im trying to test the requires of a Smart Contract using just solidity according to this article:

http://truffleframework.com/tutorials/testing-for-throws-in-solidity-tests

This is the contract, the throw proxy contract and the test:

```

/* Testing with solidity tests. */

import "truffle/Assert.sol";
import "truffle/DeployedAddresses.sol";
import "../contracts/MyContract.sol";

contract TestMyContract {

function testInitialStoredValue() {
    MyContract mycontract = new MyContract();

    uint expected = 24;

    Assert.equal(mycontract.mynumber(), expected, "First number set should be 24.");
}

function testTheThrow() {
    MyContract mycontract = new MyContract(); 

    ThrowProxy throwproxy = new ThrowProxy(address(mycontract)); 
    uint num = 7;
    MyContract(address(throwproxy)).storeNum(num);

    bool r = throwproxy.execute.gas(200000)(); 

    Assert.isFalse(r, "Should be false because is should throw!");

}

function testNoThrow() {
    MyContract mycontract = new MyContract(); 

    ThrowProxy throwproxy = new ThrowProxy(address(mycontract)); 

    MyContract(address(throwproxy)).storeNum(22);

    bool r = throwproxy.execute.gas(200000)(); 

    Assert.isTrue(r, "Should be true because is should throw!");

}

}

// Proxy contract for testing throws
contract ThrowProxy {
  address public target;
  bytes data;

  function ThrowProxy(address _target) {
    target = _target;
  }

  //prime the data using the fallback function.
  function() {
    data = msg.data;
  }

  function execute() returns (bool) {
    return target.call(data);
  }
}
```

If i run the tests i get this error:

![2](https://user-images.githubusercontent.com/14795944/41259154-f5679576-6da8-11e8-801c-784cb99d29a8.png)

If i chang the storeNum function to void from 

```
 function storeNum(uint mynum)
        public
        returns (bool success)
    {
     require(mynum > 10);
     mynumber = mynum; 
     return true;    
    }

```
to

```
 function storeNum(uint mynum)
        public
    {
     require(mynum > 10);
     mynumber = mynum; 
     return true;    
    }

```

the tests works..

Any ideas? 

Im using Truffle v4.1.11

**Comment by marcosmartinez7** [comment]
I have tried also with another functions that returns other types and the problem is the same, it only works for void functions

any ideas? 

thanks

**Comment by kyriediculous** [comment]
Also having issues with testing exceptions in JS tests currently... 

For me it's like the next test gets checked instead of the one i'm entering into my assertError.

**Comment by cgewecke** [comment]
@marcosmartinez7 This is super interesting - the ThrowProxy article was written a while ago and I wonder if recent changes at solidity have added that constraint. Will ask around this morning and see if anyone knows the answer to this question. 

Out of curiosity - do your transactional methods need return signatures? If they're used within Solidity they will automatically return `true` if they run without raising an exception and false otherwise.

**Comment by marcosmartinez7** [comment]
Hi @cgewecke , thanks for the answer. 

There is a couple of recent (aprox 5 months ago) articles and github repos using this aproach so thats an option, i dont know if it is a constraint or the proxy aproach just doesnt work now. 

I really apreciate if you can research about this! 

About the methods, the code that i have posted here is just an example, but the Smart contract that im working on is from a third party and i cannot change it, but it has the return signatures. I will ask them about that because i didnt know that now they automatically return true. 

Thanks!

**Comment by marcosmartinez7** [comment]
@cgewecke any news? The aproach that i have taken is to wrap all the functions that arent void functions in a test contract that expose that functions in a void way.. it is not something im proud of but works :/

**Comment by cgewecke** [comment]
@marcosmartinez7 I asked the @gnidan (another engineer here - he's recently written a solidity debugger and is intimately familiar with its workings :) He was surprised by this behavior.

Below is all the information I can find about `address.call` in the Soldity docs (from [here](https://solidity.readthedocs.io/en/v0.4.20/types.html?highlight=.call#members-of-addresses)):

>to interface with contracts that do not adhere to the ABI, the function call is provided which takes an arbitrary number of arguments of any type. These arguments are padded to 32 bytes and concatenated. One exception is the case where the first argument is encoded to exactly four bytes. In this case, it is not padded to allow the use of function signatures here.
```
address nameReg = 0x72ba7d8e73fe8eb666ea66babc8116a41bfb10e2;
nameReg.call("register", "MyName");
nameReg.call(bytes4(keccak256("fun(uint256)")), a);
```
> call returns a boolean indicating whether the invoked function terminated (true) or caused an EVM exception (false). It is not possible to access the actual data returned (for this we would need to know the encoding and size in advance).

There's nothing in the [release notes](https://github.com/ethereum/solidity/releases) at Solidity that make this clearer. 

The only other thing I can think of that might be related is [this thread](https://github.com/ethereum/solidity/issues/4116) at Solidity which addresses recent changes in the way return call values are handled. 

If you find anything out, please report. Will leave this issue open in the interim and thanks for the work-around, that's very helpful.

**Comment by wadelee1986** [comment]
how to resolve this problem ?


truffle(develop)> 
undefined
truffle(develop)>  Peggy.deployed().then(function(instance){return instance.newIpcERC20("test",10);});
Error: VM Exception while processing transaction: revert


function newIpcERC20(string name,uint decimals)external onlyOwner()  {//returns (address){
  require(ipcTokens[name] == address(0));
  IpcERC20  newToken= new IpcERC20(address(this),name,decimals);
  //newToken.mint(address(this),0);  // must init

  ipcTokens[name] = newToken;
  ipcTokenAddresses[newToken] = true;

  NewIpcERC20(name,newToken);
  //return newToken;
}

**Comment by schemar** [comment]
Just wanted to chime in and say that I have the same problem and am also using wrapper contracts now. It appears the problem is that the fallback function does not have a return value, right?

**Comment by rmlopes** [comment]
Sorry for bumping up this. I have been struggling with this issue for a couple of days now. It looks like as long as any modifier is used the ThrowProxy will fail as described. Using Solidity v0.5.

**Comment by sleepingevil** [comment]
I'm having the exact same issue in a JavaScript test, using truffle 5.0.9.

function I'm testing:

```
function transfer (address _to, uint256 _value) public returns (bool success) {
        require(balanceOf[msg.sender] >= _value);

        balanceOf[msg.sender] -= _value;
        balanceOf[_to] += _value;

        emit Transfer(msg.sender, _to, _value);

        return true;
    }
```

My test:

```
it('returns false for an unsuccessful transfer', () => MyToken.deployed().then(async function (instance) {
        const success = await instance.transfer.call(accounts[1], 1000001, { from: accounts[0] });

        assert.equal(success, false, 'success didn\'t equal false');
    }));
```

Expected behavior:

The test passes.

Actual behavior:

The test fails with an error thrown: `Returned error: VM Exception while processing transaction: revert`

**Comment by eggplantzzz** [comment]
@sleepingevil So that is currently the correct behavior when a require fails. So if you want to test it, you could put in a try/catch and test against the error that is thrown. So it would be something like this...
```
try {
  myContract.methodThatReverts.call();
} catch (error) {
  let result = error === isWhatIExpectItToBe;
  assert(result);
}
```
I hope that helps!

**Comment by marcosmartinez7** [comment]
@eggplantzzz No, thats not the correct behaviour in Solidity. Also, there is no try-catch block in Solidity. The code that youre posting here is a js test.

**Comment by sleepingevil** [comment]
@eggplantzzz: @marcosmartinez7 is right. This is not the expected behaviour. [ERC20 specification](https://eips.ethereum.org/EIPS/eip-20) requires the transfer method of a token to:
- Throw if the sender doesn't have sufficient balance
- Return a boolean

I'm trying to test the return value in this case, not the exception thrown.

According to the [throw testing documentation](https://truffleframework.com/tutorials/testing-for-throws-in-solidity-tests), you can do this in Solidity with ThrowProxy. You can also test it with `contract.transfer.call(...)` in javascript, which will not execute the transaction, but it will just make the method return its return value for testing purposes.

Both in Solidity and JavaScript this behavior is not as described, as the method throws a VM exception (revert).

I posted JS code here, because the expected and actual behavior are the same in both programming languages, which makes me believe that the root cause is the same. Please advise me if I should create a separate issue for the JS one (I'm just trying to avoid having duplicate issues in github).

**Comment by eggplantzzz** [comment]
@marcosmartinez7 Ah yes you are right, I did not see the Solidity and assumed it was a js test.
From the Solidity docs it states that an exception occurs when the value in a require statement evaluates to `false`. https://solidity.readthedocs.io/en/v0.5.7/control-structures.html#error-handling-assert-require-revert-and-exceptions
Am I missing something?
